OffHeapMemory, ThreadCreator duplicate classes

### Describe the bug

The build for one of our products has a check that every fully qualified classname is found in exactly one jar, with the idea that if there are multiple classes with the same name, it becomes unpredictable which one the JVM will use.

After upgrading infinispan from 15.1.5 to 15.2.1, this check triggered an error message.

We can silence this message by specifying that duplicate classes are acceptable if the bytecode is identical in all copies, or by excluding some classes from the check, but that is not ideal: silencing error messages makes it easier for other issues to slip through.

The cause is somewhere in this part of the dependency tree:
```
[INFO] |  |  +- org.infinispan:infinispan-core:jar:15.2.1.Final:compile
[INFO] |  |  |  +- org.infinispan:infinispan-commons:jar:15.2.1.Final:compile
[INFO] |  |  |  |  \- io.reactivex.rxjava3:rxjava:jar:3.1.10:compile
[INFO] |  |  |  |     \- org.reactivestreams:reactive-streams:jar:1.0.4:compile
[INFO] |  |  |  +- org.infinispan:infinispan-commons-spi:jar:15.2.1.Final:compile
```
Looking at the infinispan sources, it seems infinispan-commons does an explicit include of the commons-spi classes, and then infinispan-core depends on both commons and commons-spi, so that the application gets identical classes via two different jars.

I am not familiar enough with infinispan to offer an opinion on how to fix this.

### Expected behavior

We expect `mvn clean verify` to find no duplicate classes

### Actual behavior

```
[ERROR] Rule 0: org.codehaus.mojo.extraenforcer.dependencies.BanDuplicateClasses failed with message:
[ERROR] Duplicate classes found:
[ERROR]
[ERROR]   Found in:
[ERROR]     org.infinispan:infinispan-commons:jar:15.2.1.Final:compile
[ERROR]     org.infinispan:infinispan-commons-spi:jar:15.2.1.Final:compile
[ERROR]   Duplicate classes:
[ERROR]     org/infinispan/commons/spi/OffHeapMemory.class
[ERROR]     org/infinispan/commons/spi/ThreadCreator.class
```

### How to Reproduce?

Run `mvn clean verify` in a directory with just the following pom.xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>dupclassdemo</artifactId>
    <version>1.0.0</version>

    <dependencies>
        <dependency>
            <groupId>org.infinispan</groupId>
            <artifactId>infinispan-core</artifactId>
            <version>15.2.1.Final</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-enforcer-plugin</artifactId>
                <version>3.5.0</version>
                <dependencies>
                    <dependency>
                        <groupId>org.codehaus.mojo</groupId>
                        <artifactId>extra-enforcer-rules</artifactId>
                        <version>1.8.0</version>
                    </dependency>
                </dependencies>
                <executions>
                    <execution>
                        <id>enforce</id>
                        <goals>
                            <goal>enforce</goal>
                        </goals>
                        <configuration>
                            <rules>
                                <banDuplicateClasses>
                                    <findAllDuplicates>true</findAllDuplicates>
                                    <ignoreWhenIdentical>false</ignoreWhenIdentical>
                                </banDuplicateClasses>
                            </rules>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>

</project>
```

### Output of `uname -a` or `ver`

some kind of windows-11.

### Output of `java -version`

openjdk version "21.0.7" 2025-04-15 LTS OpenJDK Runtime Environment Temurin-21.0.7+6 (build 21.0.7+6-LTS) OpenJDK 64-Bit Server VM Temurin-21.0.7+6 (build 21.0.7+6-LTS, mixed mode, sharing)

### Infinispan version or git rev

15.2.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937) Maven home: C:\Program Files\Maven\apache-maven-3.9.9 Java version: 21.0.7, vendor: Eclipse Adoptium, runtime: C:\Program Files\Eclipse Adoptium\jdk-21.0.7.6-hotspot Default locale: nl_NL, platform encoding: UTF-8 OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"

### Additional information

_No response_

--- a/build/configuration/pom.xml
+++ b/build/configuration/pom.xml
@@ -206,6 +206,7 @@
       <version.maven.deploy>3.1.4</version.maven.deploy>
       <version.maven.enforcer>3.5.0</version.maven.enforcer>
       <version.maven.exec>3.5.0</version.maven.exec>
+      <version.maven.extra-enforcer-rules>1.10.0</version.maven.extra-enforcer-rules>
       <version.maven.git-commit-id>9.0.1</version.maven.git-commit-id>
       <version.maven.gpg>3.2.7</version.maven.gpg>
       <version.maven.help>3.5.1</version.maven.help>

--- a/build/infinispan-defaults-maven-plugin/pom.xml
+++ b/build/infinispan-defaults-maven-plugin/pom.xml
@@ -55,6 +55,33 @@
                     <skipErrorNoDescriptorsFound>true</skipErrorNoDescriptorsFound>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-enforcer-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>enforce</id>
+                        <goals>
+                            <goal>enforce</goal>
+                        </goals>
+                        <configuration>
+                            <rules>
+                                <banDuplicateClasses>
+                                    <scopes>
+                                        <scope>compile</scope>
+                                        <scope>provided</scope>
+                                    </scopes>
+                                    <ignoreClasses>
+                                        <ignoreClass>org.codehaus.plexus.util.xml.*</ignoreClass>
+                                    </ignoreClasses>
+                                    <findAllDuplicates>true</findAllDuplicates>
+                                    <ignoreWhenIdentical>false</ignoreWhenIdentical>
+                                </banDuplicateClasses>
+                            </rules>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
         </plugins>
     </build>
 </project>

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -18,10 +18,7 @@
          <groupId>org.infinispan</groupId>
          <artifactId>infinispan-commons</artifactId>
       </dependency>
-      <dependency>
-         <groupId>org.infinispan</groupId>
-         <artifactId>infinispan-commons-spi</artifactId>
-      </dependency>
+
       <dependency>
          <groupId>org.infinispan</groupId>
          <artifactId>infinispan-component-processor</artifactId>

--- a/pom.xml
+++ b/pom.xml
@@ -1882,6 +1882,13 @@
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-enforcer-plugin</artifactId>
                <version>${version.maven.enforcer}</version>
+               <dependencies>
+                  <dependency>
+                     <groupId>org.codehaus.mojo</groupId>
+                     <artifactId>extra-enforcer-rules</artifactId>
+                     <version>${version.maven.extra-enforcer-rules}</version>
+                  </dependency>
+               </dependencies>
             </plugin>
             <plugin>
                <groupId>org.apache.maven.plugins</groupId>
@@ -2315,6 +2322,30 @@
                   </goals>
                   <configuration>
                      <cpuCount>system.numCores</cpuCount>
+                  </configuration>
+               </execution>
+            </executions>
+         </plugin>
+         <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-enforcer-plugin</artifactId>
+            <executions>
+               <execution>
+                  <id>enforce</id>
+                  <goals>
+                     <goal>enforce</goal>
+                  </goals>
+                  <configuration>
+                     <rules>
+                        <banDuplicateClasses>
+                           <scopes>
+                              <scope>compile</scope>
+                              <scope>provided</scope>
+                           </scopes>
+                           <findAllDuplicates>true</findAllDuplicates>
+                           <ignoreWhenIdentical>false</ignoreWhenIdentical>
+                        </banDuplicateClasses>
+                     </rules>
                   </configuration>
                </execution>
             </executions>

--- a/quarkus/pom.xml
+++ b/quarkus/pom.xml
@@ -39,6 +39,13 @@
                     <configLocation>checkstyle-junit5.xml</configLocation>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-enforcer-plugin</artifactId>
+                <configuration>
+                    <skip>true</skip>
+                </configuration>
+            </plugin>
         </plugins>
     </build>
 

